src/libsyntax/ext/tt/macro_check.rs

```diff
@@ -0,0 +1,342 @@
+        let mut r = macros.to_owned();
+        r.push(MacroState { binders, ops });
+        r
+    };
+    let mut nested_binders = Binders::default();
+    for tt in tts {
+        match (state, tt) {
+            // . macro_rules
+            // . macro
+            (0, &TokenTree::Token(Token { kind: TokenKind::Ident(name, false), .. })) => {
+                if name == sym::macro_rules {
+                    state = 1;
+                }
+                if name == kw::Macro {
+                    state = 4;
+                }
+            }
+            // macro_rules . !
+            (1, &TokenTree::Token(Token { kind: TokenKind::Not, .. })) => {
+                state = 2;
+            }
+            // macro_rules ! . ident
+            (2, &TokenTree::Token(Token { kind: TokenKind::Ident(..), .. })) => {
+                state = 3;
+            }
+            // macro_rules ! . $meta
+            (2, &TokenTree::MetaVar(..)) => {
+                state = 3;
+                check_occurrences(sess, tt, macros, binders, ops, valid);
+            }
+            // macro_rules ! ident . {delim}
+            // macro_rules ! $meta . {delim}
+            // macro ident . {delim}
+            // macro $meta . {delim}
+            (3, &TokenTree::Delimited(_, ref del)) | (5, &TokenTree::Delimited(_, ref del))
+                if del.delim == DelimToken::Brace =>
+            {
+                state = 0;
+                let rest = check_nested_macro(sess, &del.tts, &nested_macros, valid);
+                // If we did not check the whole macro definition, then check the rest as if outside
+                // a macro definition.
+                check_nested_occurrences(sess, &del.tts[rest..], macros, binders, ops, valid);
+            }
+            // macro . ident
+            (4, &TokenTree::Token(Token { kind: TokenKind::Ident(..), .. })) => {
+                state = 5;
+            }
+            // macro . $meta
+            (4, &TokenTree::MetaVar(..)) => {
+                state = 5;
+                check_occurrences(sess, tt, macros, binders, ops, valid);
+            }
+            // macro ident . (delim)
+            // macro $meta . (delim)
+            (5, &TokenTree::Delimited(_, ref del)) if del.delim == DelimToken::Paren => {
+                state = 6;
+                nested_binders = Binders::default();
+                check_binders(sess, tt, &nested_macros, &mut nested_binders, &[], valid);
+            }
+            // macro ident (delim) . {delim}
+            // macro $meta (delim) . {delim}
+            (6, &TokenTree::Delimited(_, ref del)) if del.delim == DelimToken::Brace => {
+                state = 0;
+                check_occurrences(sess, tt, &nested_macros, &nested_binders, &[], valid);
+            }
+            (_, ref tt) => {
+                state = 0;
+                check_occurrences(sess, tt, macros, binders, ops, valid);
+            }
+        }
+    }
+}
+
+fn check_nested_macro(
```

Comment:
Please add a comment that this heuristically checks `tts` as if it was a list of `(LHS) => {RHS}` tokens. It is a best-effort to detect a macro definition. And it returns the position in `tts` where we stopped checking because we detected we were not in a macro definition any more.